File: blockfish-js/src/ipc.js
const EventEmitter = require('events');
const childProcess = require('child_process');
const jspb = require('google-protobuf');
const protos = require('../generated/blockfish_pb.js');

class IPC extends EventEmitter {
    constructor(subprocess) {
        super();
        Object.assign(this, {
            subprocess,
            _buf: Buffer.alloc(0),
            _reader: new jspb.BinaryReader,
            _killed: false,
        });
        subprocess.stdout.on('readable', this._onRead.bind(this));
        subprocess.on('exit', this._onProcessExit.bind(this));
        subprocess.on('error', e => this.emit('error', e));
    }

    _onRead() {
        if (this._killed) {
            // not sure if this can happen but in case it can, just abort
            return;
        }

        // concatenate all incoming buffers
        let bufs = [this._buf];
        let buf;
        while ((buf = this.subprocess.stdout.read()) !== null) {
            bufs.push(buf);
        }
        buf = Buffer.concat(bufs);

        // parse all (complete) responses in the buffer
        let pos = 0;
        while (pos < buf.length) {
            let resp = new protos.Response;
            let end = parse(buf, pos, this._reader, resp);
            if (end === null) {
                break;
            }
            this.emit('recv', resp);
            pos = end;
        }
        this._buf = buf.slice(pos);
    }

    _onProcessExit(sig) {
        this._killed = true;
        if (typeof sig === 'number' && sig !== 0) {
            this.emit('error', new Error("non-zero exit code " + sig));
        } else {
            this.emit('exit');
        }
    }

    send(req, cb) {
        if (this._killed) {
            return;
        }
        const chunk = req.serializeBinary();
        const lengthWriter = new jspb.BinaryWriter;
        lengthWriter.encoder_.writeUnsignedVarint32(chunk.length);
        const lengthChunk = lengthWriter.getResultBuffer();
        this.subprocess.stdin.write(lengthChunk);
        this.subprocess.stdin.write(chunk, cb);
    }

    kill() {
        if (this._killed) {
            return;
        }
        this.subprocess.kill('SIGTERM');
        this._killed = true;
        this._reader.free();
    }
}

/**
 * Returns true iff a (complete) varint was found in `buf` at `pos`.
 */
function anyVarints(buf, pos) {
    while (pos < buf.length && buf[pos] & 0x80) {
        pos++;
    }
    return pos < buf.length;
}

/**
 * Parse a response from `buf` starting at position `pos`, deserializing into `resp`. If
 * the response data is incomplete, returns `null`, otherwise returns the next position to
 * read from.
 */
function parse(buf, pos, reader, resp) {
    if (!anyVarints(buf, pos)) {
        return null;
    }

    reader.setBlock(buf, pos);
    let len = reader.decoder_.readUnsignedVarint32();
    pos = reader.getCursor();
    if (pos + len > buf.length) {
        return null;
    }

    reader.setBlock(buf, pos, len);
    protos.Response.deserializeBinaryFromReader(resp, reader);
    pos = reader.getCursor();
    return pos;
}

IPC.childProcess = (executable, args) => new IPC(
    childProcess.spawn(
        executable,
        args,
        { stdio: ['pipe', 'pipe', 'inherit'] },
    )
);

module.exports = IPC;

File: blockfish-python/blockfish/ipc.py
import asyncio, subprocess
from google.protobuf.internal.decoder import _DecodeVarint32
from google.protobuf.internal.encoder import _EncodeVarint

import blockfish.blockfish_pb2 as protos


class IPC:

    DEFAULT_BLOCKFISH_PATH = 'blockfish'

    def __init__(self, process):
        self._sp = process

    async def send(self, *reqs):
        for req in reqs:
            write_request(self._sp.stdin, req)

    async def recv(self):
        return await read_response(self._sp.stdout)

    async def kill(self):
        if self._sp.returncode is None:
            self._sp.terminate()
            await self._sp.wait()


async def create_subprocess_ipc(program = IPC.DEFAULT_BLOCKFISH_PATH):
    sp = await asyncio.create_subprocess_exec(
        program = program,
        stdout = subprocess.PIPE,
        stdin = subprocess.PIPE,
    )
    return IPC(sp)


async def read_varint(stream):
    buf = bytearray()
    bs = b''
    while not any(x & 0x80 == 0 for x in bs):
        bs = await stream.read(n = 1)
        if len(bs) == 0:
            raise EOFError
        buf += bs
    varint, pos = _DecodeVarint32(buf, 0)
    assert(pos == len(buf))
    return varint


async def read_response(stream):
    length = await read_varint(stream)
    buf = await stream.read(length)
    if len(buf) < length:
        raise EOFError
    res = protos.Response()
    res.ParseFromString(buf)
    return res


def write_request(stream, req):
    data = req.SerializeToString()
    length_buf = bytearray()
    _EncodeVarint(length_buf.extend, len(data))
    stream.write(length_buf)
    stream.write(data)
